### PR TITLE
d/lb - add `private_ipv4_address` attribute to subnet mapping

### LINKED ISSUE #14565: aws_lb Data Source returns error setting subnet_mapping: Invalid address to set: []string{"subnet_mapping", "0", "private_ipv4_address"}
_This issue was originally opened by @kartik-moolya as hashicorp/terraform#25806. It was migrated here as a result of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original body of the issue is below._
    
<hr>
    

### Terraform Version

```
Terraform v0.12.29
```

### Terraform Configuration Files

```terraform
data "aws_lb" "nginxlb" {
  arn = var.nginx_lb_arn
  name = "mvp-ui-nginx-NLB"
}
```

### Debug Output

https://gist.github.com/kartik-moolya/7f8d657aec8990eef2707565c4e77f8e

### Expected Behavior

The Data should have imported the LB details
### Actual Behavior

Got the below error
```
Error: error setting subnet_mapping: Invalid address to set: []string{"subnet_mapping", "0", "private_ipv4_address"}

  on main.tf line 39, in data "aws_lb" "nginxlb":
  39: data "aws_lb" "nginxlb" {
``` 

### Steps to Reproduce

1. `terraform init`
2. `terraform apply`
### Additional Context

The loadbalancer is a public Network load balancer and was created using cloudformation

### PR DESCRIPTION
**Pull request by DrFaust92**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Relates  #11404, Closes #14565.

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
data_source_aws_lb - add `private_ipv4_address` attribute to subnet mapping
```

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccDataSourceAWSLB_'
--- PASS: TestAccDataSourceAWSLB_basic (289.27s)
--- PASS: TestAccDataSourceAWSLB_BackwardsCompatibility(301.30s)
```

### CODE REVIEW

**Comment by bflad** [comment]
Sigh, #8223 rears its head again. 😖  Thanks for working on this, @DrFaust92, is this still WIP (according to the PR title)?

**Comment by DrFaust92** [comment]
@bflad, yeah debugging an issue where subnet_mapping differs from data source and resource. Looking into it.

**Comment by DrFaust92** [comment]
@bflad, there was another issue with the Set func missing hiding there as well.

**Comment by ewbankkit** [comment]
Closes #14565.

**Comment by kartik-moolya** [comment]
Can we expedite this please. I'm having to move all of my code to cloudformation due to this bug :(

File: aws/data_source_aws_lb.go [hunk 1]
```diff
@@ -1,11 +1,13 @@
 package aws
 
 import (
+	"bytes"
 	"fmt"
 	"log"
 
 	"github.com/aws/aws-sdk-go/aws"
 	"github.com/aws/aws-sdk-go/service/elbv2"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/hashcode"
```
Comment (bflad) [inline aws/data_source_aws_lb.go:10]: This import was very recently moved as part of our Terraform Plugin SDK v1 removal: #14593

It now lives at:

```go
"github.com/terraform-providers/terraform-provider-aws/aws/internal/hashcode"
```

That being said, adding `Set` functions to data sources should generally not be necessary and especially after this fix in SDK version 2: https://github.com/hashicorp/terraform-plugin-sdk/pull/197

If this is rebased against the current commit on the main branch, which is all up to date with SDK v2, the `Set` function can hopefully be removed.

File: aws/data_source_aws_lb.go [hunk 2]
```diff
@@ -61,14 +63,30 @@ func dataSourceAwsLb() *schema.Resource {
...
+					m := v.(map[string]interface{})
+					buf.WriteString(fmt.Sprintf("%s-", m["subnet_id"].(string)))
+					if m["allocation_id"] != "" {
+						buf.WriteString(fmt.Sprintf("%s-", m["allocation_id"].(string)))
+					}
+					if m["private_ipv4_address"] != "" {
+						buf.WriteString(fmt.Sprintf("%s-", m["private_ipv4_address"].(string)))
+					}
+					return hashcode.String(buf.String())
+				},
```
Comment (bflad) [inline aws/data_source_aws_lb.go:89]: See note above 👍 

```suggestion
```

File: aws/data_source_aws_lb.go [hunk 3]
```diff
@@ -61,14 +63,30 @@ func dataSourceAwsLb() *schema.Resource {
 					Schema: map[string]*schema.Schema{
 						"subnet_id": {
 							Type:     schema.TypeString,
-							Required: true,
```
Comment (bflad) [inline aws/data_source_aws_lb.go:64]: Thank you for fixing these. 😄

**Comment by DrFaust92** [comment]
Thanks @bflad, fixed and added some more tests to cover all attributes

**Pull request merged**